### ISSUE TITLE
Incoming is not well suited for "immutable" objects

### ISSUE THREAD

**Issue by castarco** [Rican7/incoming#4]
Hi,

I think that the interfaces should be changed to allow constructing objects only by calling its constructors, without the need of using setter methods.

In the exposed examples (README) we can see that the object is constructed before being filled, this imposes several restrictions on the software architecture, because it makes impossible to combine this library with immutable objects.

**Comment by thepsion5** [comment]
To add to this, I would be happy to submit a PR allowing for entities to be constructed rather than passed in, but I'd like the author's input on how to go about doing so.

Should this be separate from the Hydrator interface (since hydration implies adding data in an existing model)? Or should there be an additional interface that extends Hydrator and adds a `factory($input)` method? Both approaches have drawbacks.

**Comment by Rican7** [comment]
I **definitely** see where you're coming from. Immutable value-objects are definitely ideal in many scenarios. I'm personally a big fan of them. In fact, the structures provided in this lib ([`FixedList`][FixedList], [`Map`][Map], etc) are immutable.

Many immutable object APIs contain immutable update methods, for example [the PSR-7 `HttpMessage::withHeader()` method][psr7-httpmessage-withheader]. I do, however, understand that such an API isn't always ideal or possible. This reason, however, is one of the main factors as to why [the hydrators are designed to **return** the object that is being hydrated][hydratorinterface-returndoc]. This design allows more freedom with the hydrators (otherwise they'd return nothing/`void` and simply modify the incoming object). To work with immutable objects, then, you can choose to do one of a few things:

- Provide a `withX()` style API for your objects. _(Not always an option or desirable API)_
- Inject a [**factory**][wiki-factory-pattern] into the hydrator, and have the factory do the building via delegation.
- Provide the immutable object to the hydrator and then simply build another object in the hydrator, like so:
```php
class User
{
    private $id;
    private $name;

    public function __construct($id, $name)
    {
        $this->id = $id;
        $this->name = $name;
    }

    public function getId()
    {
        return $this->id;
    }
}

class UserHydrator implements Incoming\Hydrator\HydratorInterface
{
    public function hydrate($input, User $user)
    {
        return new User(
            $user->getId(), // Get the existing user ID, as we want that to stay the same
            $input['name'])
        );
    }
}
```

I think these options provide enough ability to use **Incoming** and it's hydrators and processor to support Immutable objects without opinionating the library or forcing it's users into a specific pattern.

I hope this response is helpful. Let me know if there's something that can be clarified, or if you have any ideas for improving the current situation beyond the suggestions that I've made.

If this was helpful enough, I may consider adding a section on "Working With Immutable Objects" to the README or Wiki. 😃 




[FixedList]: https://github.com/Rican7/incoming/blob/v0.2.1/src/Incoming/Structure/FixedList.php
[Map]: https://github.com/Rican7/incoming/blob/v0.2.1/src/Incoming/Structure/Map.php
[psr7-httpmessage-withheader]: https://github.com/php-fig/http-message/blob/1.0/src/MessageInterface.php#L117-L132
[hydratorinterface-returndoc]: https://github.com/Rican7/incoming/blob/v0.2.1/src/Incoming/Hydrator/HydratorInterface.php#L27
[wiki-factory-pattern]: https://en.wikipedia.org/wiki/Factory_(object-oriented_programming)

**Comment by thepsion5** [comment]
It just now occurred to me that I should be able to just pass null as the second argument as `process()` which is good enough for me personally. An addition to the readme wouldn't hurt though!

**Comment by shadowhand** [comment]
@Rican7 that's super helpful. It would definitely help add some documentation for those of us that use immutable objects.

**Comment by shadowhand** [comment]
After looking through this again, I realized that it would be possible to:

- Create a factory that uses the `::class` of an immutable object
- Create a hydrator that takes the `::class` as the model and ignores it

Basically:

```php
class ImmutableFactory implements HydratorFactoryInterface
{
    const MAPPING = [
        User::class => UserHydrator::class,
        // ...
   ];

    public function buildForModel($model)
    {
        return new self::MAPPING[$model]();
    }
}
```

```php
class UserHydrator implements HydratorInterface
{
    public function hydrate(array $data, $model)
    {
        // ignore $model completely, or use: new $model();
        return new User(
            $data['username'],
            // ...
        );
    }
}
```

And finally:

```php
$user = $incoming->process($_POST, User::class);
```

**Comment by Rican7** [comment]
@shadowhand Yea, that's definitely one way of doing it! If your hydrators have dependencies that they expect to be injected then it gets a bit more complex, but yea good strategy!

**Comment by Rican7** [comment]
Boom! The concept of **"Builders"** has been introduced in the newly released [**v0.3.0**](https://github.com/Rican7/incoming/releases/tag/v0.3.0)!

Check it out and let me know what you think! 😃
